### DISCUSSION TITLE
Getting to function definition outside my project

### DISCUSSION THREAD

**Discussion by GrqPnegre** [clangd/clangd#1241] [Q&A]
I'm confused how clangd is able to locate function definitions outside my project.  I've got something working, but it's magic to me.  My project depends on a static library, for which I built a ```compile_commands.json```.  The build directory and source directory sit at the same level:

```
work/bar
work/bar-outdir
```

The library is then installed into a separate directory

```
.local/bar
.local/bar/include
.local/bar/lib
```
My application uses ```bar``` can get to bar's function definitions by first going though the header ```.local/bar/include/bar.h``` and then to ```work/bar/src/bar.c```.  This works because the json file is located in ```work/compile_commands.json```, and a ```.clangd``` file located in ```.local/bar``` pointing to that json.  Once ```bar.c``` has been open, subsequent queries go directly the the function definition in ```bar.c```.

If I leave ```compile_commands.json``` in ```work/bar-outdir```, then clangd isn't able to find bar's definitions.  I expected that the file locations specified in ```compile_commands.json``` would be enough to build the right index, and that ```compile_commands.json``` could then be located anywhere.  This isn't the case.  This is the magically part.  It looks like clangd is searching child dirs below the json file.

My questions are
1. Why does clangd need ```compile_commands.json``` to be a parent of the source?  What's going on there?
2. How does clangd find function definitions outside a project when using ```compile_commands.json```?
3. What the recommended approach for handling external translation units?

**Comment by HighCommander4** [comment]
You might find https://clangd.llvm.org/faq#how-do-i-fix-errors-i-get-when-opening-headers-outside-of-my-project-directory helpful

**Reply by GrqPnegre** [reply]
Thx for the reference.

After reading over the FAQ and https://clangd.llvm.org/design/indexing I think I understand what's going on.  When the first header file reference is made ```clangd``` tries to locate the CDB relative to that location (and parents).  In my case it picks up the ```.clangd``` project config which uses the key ```CompilationDatabase```, loads the CDB and creates a relative ```.cache``` of all sources from that CDB.  The subsequent _find definition_ then can locate the definition from the header file's declaration.  Not magic anymore.

I did mistakenly concatenate the CDBs using ```cat``` instead of using something appropriate like ```jq```.

I wonder if it'd be helpful for ```clangd``` and ```.clangd``` to take multiple CDBs to express dependencies.  

```yaml
#File: .clangd
CompileFlags:
  CompilationDatabase: /path/to/cdb1
  CompilationDatabase: /path/to/cdb2

CompileFlags:
  CompilationDatabase: /path/to/cdb3
```

Also, might be interesting to consider the use-case where sources and the CDB are bundled similarly to something like debug symbols a la MSFT.  That way, if someone needs to generate a cache for a subset of dependencies, they can pull it down and just reference the install location.

Having now played with ```lsp-mode``` in Emacs with ```clangd```, this is really killer.  I liked my other setup with ```GNU globals```, but this is next level.

**Reply by HighCommander4** [reply]
Glad you were able to get it sorted!

> I wonder if it'd be helpful for `clangd` and `.clangd` to take multiple CDBs to express dependencies.

There's some discussion of a similar request in #1011
